File: lua/meetup.lua
local wday = {
    Sunday = 1,
    Monday = 2,
    Tuesday = 3,
    Wednesday = 4,
    Thursday = 5,
    Friday = 6,
    Saturday = 7,
}

local index = {
    first = 1,
    second = 2,
    third = 3,
    fourth = 4,
    fifth = 5,
}

return function(config)
    local t1 = os.time({
        year = config.year,
        month = config.month,
        day = 1,
    })

    local d = os.date("*t", t1)
    d.month = d.month + 1
    d.day = d.day - 1

    local t2 = os.time(d)

    local arr = {}
    local one_day = 60 * 60 * 24
    for t = t1, t2, one_day do
        local d = os.date("*t", t)
        if d.wday == wday[config.day] then
            table.insert(arr, d.day)
        end
    end
    if config.week == "teenth" then
        for _, day in ipairs(arr) do
            if day >= 13 and day <= 19 then
                return day
            end
        end
    end

    index.last = #arr
    return arr[index[config.week]]
end

File: elixir/meetup.ex
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @type weekday ::
          :monday
          | :tuesday
          | :wednesday
          | :thursday
          | :friday
          | :saturday
          | :sunday

  @weekdays [:monday, :tuesday, :wednesday, :thursday, :friday, :saturday, :sunday]

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: Date.t()
  def meetup(year, month, weekday, schedule) do
    n = Enum.find_index(@weekdays, &(&1 == weekday)) + 1

    Date.new!(year, month, 1)
    |> Stream.unfold(fn
      dt = %Date{month: m} when m == month -> {dt, Date.add(dt, 1)}
      _ -> nil
    end)
    |> Stream.filter(fn dt -> Date.day_of_week(dt) == n end)
    |> then(fn dates ->
      case schedule do
        :first -> Enum.at(dates, 0)
        :second -> Enum.at(dates, 1)
        :third -> Enum.at(dates, 2)
        :fourth -> Enum.at(dates, 3)
        :last -> Enum.to_list(dates) |> List.last()
        _ -> Enum.find(dates, fn dt -> dt.day in 13..19 end)
      end
    end)
  end
end
